### PR TITLE
Add slicing support for fbgemm fp8 and int4

### PR DESCRIPTION
**Pull request by jerryzh168**
Summary:
att, this is needed in vllm
Note that irregular shapes will require padding, which is not implemented right now, we can add that if it's required by the model

Test Plan:
```
python test/dtypes/test_fbgemm_fp8.py -k test_slice 
python test/dtypes/test_fbgemm_int4.py -k test_slice
```

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: test/dtypes/test_fbgemm_fp8.py [hunk 1]
```diff
@@ -34,15 +41,43 @@ def test_linear(self):
...
+        weight1 = dummy.weight.narrow(0, 0, 64)
+        weight2 = dummy.weight.narrow(1, 0, 128)
+
+        # check for sliced weight, before and after float8 quantization
+        # does not differ too much
+        input = torch.randn(2, 256, dtype=dtype, device=device)
+        res_ref = dummy1(input)
+        dummy.weight = torch.nn.Parameter(weight1, requires_grad=False)
+        res = dummy(input)
+        assert compute_error(res, res_ref) > 25
```
Comment (drisspg) [inline test/dtypes/test_fbgemm_fp8.py:77]: can this be made more concrete, e.g. we can test the exact elements match

File: test/dtypes/test_fbgemm_int4.py [hunk 2]
```diff
@@ -33,16 +41,43 @@ def test_linear(self):
...
+        dummy1.weight = torch.nn.Parameter(
+            dummy.weight.narrow(0, 0, 64), requires_grad=False
+        )
+        dummy2 = torch.nn.Linear(128, 256, dtype=dtype, device=device)
+        dummy2.weight = torch.nn.Parameter(
+            dummy.weight.narrow(1, 0, 128), requires_grad=False
+        )
+
+        quantize_(dummy, self.config)
+        weight1 = dummy.weight.narrow(0, 0, 64)
```
Comment (drisspg) [inline test/dtypes/test_fbgemm_int4.py:62]: ditto here

File: torchao/dtypes/fbgemm_fp8_tensor.py [hunk 3]
```diff
@@ -146,6 +155,43 @@ def _(func, types, args, kwargs):
     )
 
 
+@implements(aten.slice.Tensor)
+def _(func, types, args, kwargs):
+    self, dim, start, end, step = fill_defaults(args, 5, [0, None, None, 1])
```
Comment (drisspg) [inline torchao/dtypes/fbgemm_fp8_tensor.py:198]: I think this stuff is error prone, do you think that : https://github.com/pytorch/ao/blob/488ecd4cf871035b091981a557a9204a5d50bf5b/torchao/dtypes/floatx/float8_layout.py#L302

is helpful here?

File: test/dtypes/test_fbgemm_int4.py (hunk 2 above)
Comment (jerryzh168) [reply test/dtypes/test_fbgemm_int4.py:62]: we can have exact match if dequantize op is implemented I think, right now it's just a sanity check, can we follow up this later?

File: torchao/dtypes/fbgemm_fp8_tensor.py (hunk 3 above)
Comment (jerryzh168) [reply torchao/dtypes/fbgemm_fp8_tensor.py:198]: the scale here is not 2d actually, but might be useful if we resize that to 2D/3D

Comment (drisspg) [reply torchao/dtypes/fbgemm_fp8_tensor.py:198]: I dont think this relies on 2d vs 3d no?

File: test/dtypes/test_fbgemm_int4.py (hunk 2 above)
Comment (drisspg) [reply test/dtypes/test_fbgemm_int4.py:62]: Can you ensure that the weight1 's data + its scale is exactly the right slice of dummy.weight

Comment (jerryzh168) [reply test/dtypes/test_fbgemm_int4.py:62]: yeah I can add an assert for that to check the internal data

File: torchao/dtypes/fbgemm_fp8_tensor.py (hunk 3 above)
Comment (jerryzh168) [reply torchao/dtypes/fbgemm_fp8_tensor.py:198]: does this assume scale has the same shape as data: https://github.com/pytorch/ao/blob/488ecd4cf871035b091981a557a9204a5d50bf5b/torchao/dtypes/floatx/float8_layout.py#L321

Comment (drisspg) [reply torchao/dtypes/fbgemm_fp8_tensor.py:198]: hmm yeah it does, we should probably assert

File: test/dtypes/test_fbgemm_fp8.py (hunk 1 above)
Comment (jerryzh168) [reply test/dtypes/test_fbgemm_fp8.py:77]: updated above, please take a look again

File: torchao/dtypes/fbgemm_int4_tensor.py [hunk 4]
```diff
@@ -151,13 +153,96 @@ def _(func, types, args, kwargs):
...
+        for tensor_name in self_tensors:
+            getattr(self, tensor_name).copy_(getattr(src, tensor_name))
+        return
+    raise ValueError(
+        f"Not supported args for copy_ due to metadata mismatch: {args[0], args[1]}"
+    )
+
+
+@implements(aten.slice.Tensor)
+def _(func, types, args, kwargs):
```
Comment (drisspg) [inline torchao/dtypes/fbgemm_int4_tensor.py:190]: Maybe a doc block here on what is supported and the semantics this code is not that easy to grok

**Review by drisspg — APPROVED** [review]
